File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <bits/stdc++.h>

using namespace std;

bool isDiffOneWord(string a, string b){
    int count = 0;
    for(int i = 0; i < a.length(); i++){
        if(a[i]==b[i]) continue;
        count++;
        if(count > 1) return false;
    }
    return true;
}

int solution(string begin, string target, vector<string> words) {
    int answer = 0;
    map<string, bool>visited;
    for(string word : words){
        visited[word] = false;
    }
    if(visited.find(target) == visited.end()) return 0;
    queue<pair<string, int>>q;
    q.push({begin, 0});
    while(!q.empty()){
        string tmpString = q.front().first;
        int tmpCount = q.front().second;
        q.pop();
        if(tmpString == target) return tmpCount;
        if(visited[tmpString]) continue;
        visited[tmpString] = true;
        for(auto word : words){
            if(visited[word] == false && isDiffOneWord(tmpString, word)){
                q.push({word, tmpCount+1});
            }
        }
    }
    
    return 0;
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;
import java.util.stream.*;

class Solution {
    
    boolean isOneWordDiff(String a, String b){
        return 
            IntStream.range(0, a.length())
            .filter(i -> a.charAt(i) != b.charAt(i))
            .count() == 1 ? true : false;
    }
    
    class Data{
        String word;
        int count;
        Data(String word, int count){
            this.word = word;
            this.count = count;
        }
    }
    
    public int solution(String begin, String target, String[] words) {
        Set<String> set = new HashSet<>();
        Queue<Data> q = new LinkedList<>();
        q.add(new Data(begin, 0));
        
        while(!q.isEmpty()){
            Data data = q.poll();
            String word = data.word;
            int count = data.count;
            
            if(set.contains(word)) continue;
            set.add(word);
            if(word.equals(target)) return count;
            
            for(String nextWord : words){
                if(isOneWordDiff(word, nextWord)) q.add(new Data(nextWord, count+1));
            }
        }
        return 0;
    }
}